hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/builtin/SqlPageCodec.java

```diff
@@ -0,0 +1,181 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.impl.protocol.codec.builtin;
+
+import com.hazelcast.client.impl.protocol.ClientMessage;
+import com.hazelcast.internal.serialization.Data;
+import com.hazelcast.sql.SqlColumnType;
+import com.hazelcast.sql.impl.client.SqlPage;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.hazelcast.client.impl.protocol.ClientMessage.BEGIN_FRAME;
+import static com.hazelcast.client.impl.protocol.ClientMessage.END_FRAME;
+import static com.hazelcast.client.impl.protocol.codec.builtin.CodecUtil.fastForwardToEndFrame;
+
+public final class SqlPageCodec {
+
+    private SqlPageCodec() {
+    }
+
+    @SuppressWarnings({"unchecked", "checkstyle:CyclomaticComplexity"})
+    public static void encode(ClientMessage clientMessage, SqlPage sqlPage) {
+        clientMessage.add(BEGIN_FRAME.copy());
+
+        // Write the "last" flag.
+        byte[] content = new byte[]{(byte) (sqlPage.isLast() ? 1 : 0)};
+        clientMessage.add(new ClientMessage.Frame(content));
+
+        // Write column types.
+        List<SqlColumnType> columnTypes = sqlPage.getColumnTypes();
+        List<Integer> columnTypeIds = new ArrayList<>(columnTypes.size());
+
+        for (SqlColumnType columnType : columnTypes) {
+            columnTypeIds.add(columnType.getId());
+        }
+
+        ListIntegerCodec.encode(clientMessage, columnTypeIds);
+
+        // Write columns.
+        for (int i = 0; i < sqlPage.getColumnCount(); i++) {
+            SqlColumnType columnType = columnTypes.get(i);
+            Iterable<?> column = sqlPage.getColumnValuesForServer(i);
+
+            switch (columnType) {
+                case VARCHAR:
+                    ListMultiFrameCodec.encodeContainsNullable(clientMessage, (Iterable<String>) column, StringCodec::encode);
+
+                    break;
+
+                case BOOLEAN:
+                    ListCNBooleanCodec.encode(clientMessage, (Iterable<Boolean>) column);
+
+                    break;
+
+                case TINYINT:
+                    ListCNByteCodec.encode(clientMessage, (Iterable<Byte>) column);
+
+                    break;
+
+                case INTEGER:
+                    ListCNIntegerCodec.encode(clientMessage, (Iterable<Integer>) column);
+
+                    break;
+
+                case BIGINT:
+                    ListCNLongCodec.encode(clientMessage, (Iterable<Long>) column);
+
+                    break;
+
+                case SMALLINT:
```

Comment:
Can you add a marker here to make sure they are fixed before 4.2? At least a comment, assert, warning log, something.

Reply:
Fixed in the commit https://github.com/hazelcast/hazelcast/pull/18089/commits/44d86df15c3ac1aedd1c3280e5fe04becfb3d4c3